File: source/problems/leetcode/62_unique_path.py
# https://leetcode.com/problems/unique-paths/

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        if m < 1 or n < 1:
            return 0
        
        dp = [[0] * (n+1) for _ in range(m+1)]
        
        for i in range(1,m+1):
            for j in range(1, n+1):
                if i == 1 and j == 1:
                    dp[i][j] = 1
                else:
                    dp[i][j] = dp[i-1][j] + dp[i][j-1]
        
        return dp[m][n]
    

    

File: source/problems/leetcode/62_unique_path.swift
// https://leetcode.com/problems/unique-paths/description/
// time: O(m*n)
// space: O(n)
class Solution {
    func uniquePaths(_ m: Int, _ n: Int) -> Int {
        var dp = Array(repeating: 1, count: n)

        for i in 1..<m {
            for j in 0..<n {
                if j == 0 {
                    dp[j] = 1
                } else {
                    dp[j] = dp[j-1] + dp[j]
                }    
            }
        }

        return dp[n-1]
    }
}